samplers/jaegerremote/sampler.go

```diff
@@ -0,0 +1,338 @@
+	if s.rateLimiter == nil {
+		s.rateLimiter = utils.NewRateLimiter(maxTracesPerSecond, math.Max(maxTracesPerSecond, 1.0))
+	} else {
+		s.rateLimiter.Update(maxTracesPerSecond, math.Max(maxTracesPerSecond, 1.0))
+	}
+	s.maxTracesPerSecond = maxTracesPerSecond
+	return s
+}
+
+func (s *rateLimitingSampler) ShouldSample(p trace.SamplingParameters) trace.SamplingResult {
+	psc := oteltrace.SpanContextFromContext(p.ParentContext)
+	if s.rateLimiter.CheckCredit(1.0) {
+		return trace.SamplingResult{
+			Decision:   trace.RecordAndSample,
+			Tracestate: psc.TraceState(),
+		}
+	}
+	return trace.SamplingResult{
+		Decision:   trace.Drop,
+		Tracestate: psc.TraceState(),
+	}
+}
+
+// Update reconfigures the rate limiter, while preserving its accumulated balance.
+// Locking must be done externally.
+func (s *rateLimitingSampler) Update(maxTracesPerSecond float64) {
+	if s.maxTracesPerSecond != maxTracesPerSecond {
+		s.init(maxTracesPerSecond)
+	}
+}
+
+// Equal compares with another sampler.
+func (s *rateLimitingSampler) Equal(other trace.Sampler) bool {
+	if o, ok := other.(*rateLimitingSampler); ok {
+		return s.maxTracesPerSecond == o.maxTracesPerSecond
+	}
+	return false
+}
+
+func (s *rateLimitingSampler) Description() string {
+	return "rateLimitingSampler{}"
+}
+
+// -----------------------
+
+// guaranteedThroughputProbabilisticSampler is a sampler that leverages both probabilisticSampler and
+// rateLimitingSampler. The rateLimitingSampler is used as a guaranteed lower bound sampler such that
+// every operation is sampled at least once in a time interval defined by the lowerBound. ie a lowerBound
+// of 1.0 / (60 * 10) will sample an operation at least once every 10 minutes.
+//
+// The probabilisticSampler is given higher priority when tags are emitted, ie. if IsSampled() for both
+// samplers return true, the tags for probabilisticSampler will be used.
+type guaranteedThroughputProbabilisticSampler struct {
+	probabilisticSampler *probabilisticSampler
+	lowerBoundSampler    *rateLimitingSampler
+	samplingRate         float64
+	lowerBound           float64
+}
+
+func newGuaranteedThroughputProbabilisticSampler(lowerBound, samplingRate float64) *guaranteedThroughputProbabilisticSampler {
+	s := &guaranteedThroughputProbabilisticSampler{
+		lowerBoundSampler: newRateLimitingSampler(lowerBound),
+		lowerBound:        lowerBound,
+	}
+	s.setProbabilisticSampler(samplingRate)
+	return s
+}
+
+func (s *guaranteedThroughputProbabilisticSampler) setProbabilisticSampler(samplingRate float64) {
+	if s.probabilisticSampler == nil {
+		s.probabilisticSampler = newProbabilisticSampler(samplingRate)
+	} else if s.samplingRate != samplingRate {
+		s.probabilisticSampler.init(samplingRate)
+	}
+	// since we don't validate samplingRate, sampler may have clamped it to [0, 1] interval
+	samplingRate = s.probabilisticSampler.SamplingRate()
+	if s.samplingRate != samplingRate {
+		s.samplingRate = s.probabilisticSampler.SamplingRate()
+	}
```

Comment:
```suggestion
	s.samplingRate = s.probabilisticSampler.SamplingRate()
```

Is there a reason to not do this unconditionally?